src/View/Helper/PaginatorHelper.php

```diff
@@ -531,39 +536,54 @@ public function generateUrlParams(array $options = [], ?string $model = null): a
         $paging = $this->params($model);
         $paging += ['page' => null, 'sort' => null, 'direction' => null, 'limit' => null];
 
-        if (!empty($paging['sort']) && !empty($options['sort']) && strpos($options['sort'], '.') === false) {
-            $paging['sort'] = $this->_removeAlias($paging['sort'], null);
+        if (!empty($paging['sort'])
+            && !empty($options['?']['sort'])
+            && strpos($options['?']['sort'], '.') === false
+        ) {
+            $paging['sort'] = $this->_removeAlias($paging['sort'], $model = null);
         }
-        if (!empty($paging['sortDefault']) && !empty($options['sort']) && strpos($options['sort'], '.') === false) {
+        if (!empty($paging['sortDefault'])
+            && !empty($options['?']['sort'])
+            && strpos($options['?']['sort'], '.') === false
+        ) {
             $paging['sortDefault'] = $this->_removeAlias($paging['sortDefault'], $model);
         }
 
-        $url = [
-            'page' => $paging['page'],
-            'limit' => $paging['limit'],
-            'sort' => $paging['sort'],
-            'direction' => $paging['direction'],
-        ];
+        $url = [];
 
         if (!empty($this->_config['options']['url'])) {
             $key = implode('.', array_filter(['options.url', Hash::get($paging, 'scope', null)]));
-            $url = array_merge($url, Hash::get($this->_config, $key, []));
+            $url = Hash::merge($url, Hash::get($this->_config, $key, []));
         }
 
         $url = array_filter($url, function ($value) {
             return $value || is_numeric($value) || $value === false;
         });
-        $url = array_merge($url, $options);
 
-        if (!empty($url['page']) && $url['page'] === 1) {
-            $url['page'] = false;
+        if (!isset($url['?'])) {
+            $url['?'] = [];
         }
+        $url['?'] += [
+            'page' => $paging['page'],
+            'limit' => $paging['limit'],
+            'sort' => $paging['sort'],
+            'direction' => $paging['direction'],
+        ];
+
+        $url['?'] = array_filter($url['?'], function ($value) {
+            return $value || is_numeric($value) || $value === false;
+        });
 
-        if (isset($paging['sortDefault'], $paging['directionDefault'], $url['sort'], $url['direction']) &&
-            $url['sort'] === $paging['sortDefault'] &&
-            strtolower($url['direction']) === strtolower($paging['directionDefault'])
+        $url = Hash::merge($url, $options);
+
+        if (!empty($url['?']['page']) && $url['?']['page'] === 1) {
+            $url['?']['page'] = null;
+        }
+        if (isset($paging['sortDefault'], $paging['directionDefault'], $url['?']['sort'], $url['?']['direction']) &&
+            $url['?']['sort'] === $paging['sortDefault'] &&
```

Comment:
Given that we have a limited number of pagination parameters would it be possible to shift only those keys into `?` in `generateUrl`? Or do scopes make that complicated?

Reply:
The scoping feature surely makes things difficult. It adds too much complexity for a feature needed/used by a few.

Reply:
It might help to have separate arguments in `generateUrl()` for the URL and paging params.

Reply:
That sounds like a great idea. :+1: